TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php

```diff
@@ -145,70 +146,137 @@ public function setName($newName) {
-			/** @var NodeData $existingNodeData */
-			foreach ($existingNodeDataArray as $existingNodeData) {
-				if ($existingNodeData->getIdentifier() !== $this->getIdentifier()) {
-					throw new NodeException(sprintf('Can not rename the node "%s" as a node already exists on path "%s"', $this->getPath(), $path), 1414436551);
-				}
-			}
+		$pathAvailable = $checkForExistence ? $this->isNodePathAvailable($path) : TRUE;
+		if (!$pathAvailable) {
+			throw new NodeException(sprintf('Can not rename the node "%s" as a node already exists on path "%s"', $this->getPath(), $path), 1414436551);
 		}
 
-		$changedNodePathsCollection = array();
+		$changedNodePathsCollection = $this->setPathInternal($path, !$checkForExistence);
+		$this->nodeDataRepository->persistEntities();
+		array_walk($changedNodePathsCollection, function($changedNodePathInformation) { call_user_func_array([$this, 'emitNodePathChanged'], $changedNodePathInformation); });
+	}
+
+	/**
+	 * Checks if the given node path is available for this node, so either no node with this path exists or an existing node has the same identifier.
+	 *
+	 * @param string $path
+	 * @return boolean
+	 */
+	protected function isNodePathAvailable($path) {
+		$existingNodeDataArray = $this->nodeDataRepository->findByPathWithoutReduce($path, $this->context->getWorkspace());
+
+		$nonMatchingNodeData = array_filter($existingNodeDataArray, function(NodeData $nodeData) {
+			return ($nodeData->getIdentifier() !== $this->getIdentifier());
+		});
 
+		return ($nonMatchingNodeData === []);
+	}
+
+	/**
+	 * Moves a node and sub nodes to the new path.
+	 * This process is different depending on the fact if the node is an aggregate type or not.
+	 *
+	 * @param string $destinationPath the new node path
+	 * @param boolean $recursiveCall is this a recursive call
+	 * @return array NodeVariants and old and new paths
+	 */
+	protected function setPathInternal($destinationPath, $recursiveCall) {
 		if ($this->getNodeType()->isAggregate()) {
+			return $this->setPathInternalForAggregate($destinationPath, $recursiveCall);
+		}
 
-			$nodeDataVariantsAndChildren = $this->nodeDataRepository->findByPathWithoutReduce($originalPath, $this->context->getWorkspace(), TRUE, TRUE);
-
-			/** @var NodeData $nodeData */
-			foreach ($nodeDataVariantsAndChildren as $nodeData) {
-				$nodeVariant = $this->createNodeForVariant($nodeData);
-				if ($nodeVariant !== NULL) {
-					$relativePathSegment = NodePaths::getRelativePathBetween($originalPath, $nodeVariant->getPath());
-					$newNodeVariantPath = NodePaths::addNodePathSegment($path, $relativePathSegment);
-					$possibleShadowedNodeData = $nodeData->move($newNodeVariantPath, $this->context->getWorkspace());
-					$nodeVariant->setNodeData($possibleShadowedNodeData);
-					$changedNodePathsCollection[] = array($nodeVariant, $originalPath, $nodeVariant->getNodeData()->getPath(), !$checkForExistence);
```

Comment:
I don't get what you mean by that exactly, but I would suggest to separate refactoring from modified behavior to make this one easier to review & test